src/autopas/utils/SimilarityFunctions.h

```diff
@@ -8,106 +8,104 @@
 
 #include "ThreeDimensionalMapping.h"
 #include "autopas/containers/ParticleContainerInterface.h"
+#include "autopas/utils/ArrayMath.h"
 #include "autopas/utils/WrapMPI.h"
 
 namespace autopas::utils {
 
 /**
- * Calculates homogeneity and max density of given AutoPas simulation.
+ * Calculates homogeneity and max density of given AutoPas container.
  * Both values are computed at once to avoid iterating over the same space twice.
  * homogeneity > 0.0, normally < 1.0, but for extreme scenarios > 1.0
- * maxDensity > 0.0, normally < 3.0, but for extreme scenarios >> 3.0
+ * maxDensity > 0.0, normally < 3.0, but for extreme scenarios > 3.0
+ *
+ * These values are calculated by dividing the container's domain into bins of perfectly equal cuboid shapes.
+ *
+ * @note The bin shapes between different AutoPas containers will not match if the dimensions of their domains don't
+ * match. Bins with greater surface area are probably more likely to feature fluctuations in the density.
+ *
+ * Not a rigorous proof, but should give an idea:
+ * - Assume homogeneous distribution.
+ * - For every particle assume it could move in any direction with equal probability (not a terrible assumption).
+ * - So every particle near a bin face has a chance to move into the neighboring bin.
+ * - And this probability is independent of the bin (i.e. the bin shape)
+ * - With a larger surface area, there are more particles with a probability of moving into the neighboring bin.
+ * - On average, same density, but each individual bin has a greater probability of losing or gaining a lot of
+ * particles.
+ *
+ * This is probably still fine for MPI Parallelized Tuning purposes, but be careful with comparing homogeneities and
+ * densities.
  *
  * @tparam Particle
  * @param container container of current simulation
- * @param startCorner lower left front corner of the box where the homogeneity shall be calculated.
- * @param endCorner upper right back corner of the box where the homogeneity shall be calculated.
  * @return {homogeneity, maxDensity}
  */
-template <class Container>
-std::pair<double, double> calculateHomogeneityAndMaxDensity(const Container &container,
-                                                            const std::array<double, 3> &startCorner,
-                                                            const std::array<double, 3> &endCorner) {
-  unsigned int numberOfParticles = static_cast<unsigned int>(container.getNumberOfParticles());
-  autopas::AutoPas_MPI_Allreduce(AUTOPAS_MPI_IN_PLACE, &numberOfParticles, 1, AUTOPAS_MPI_UNSIGNED_INT, AUTOPAS_MPI_SUM,
-                                 AUTOPAS_MPI_COMM_WORLD);
```

Comment:
I noticed that we have removed this `MPI_AllReduce` call in this refactoring. I was looking through it and it seems like this might be needed by autotuner when MPI is on, even for md-flexible simulations, but I am not sure. 

`homogenity` and `maxDensity` is used by `MPIParallelizedStrategy` class and earlier it was getting the values calculated over the entire domain and not just individual rank (which is the case now). @FG-TUM probably can tell us better about this. We might need to adjust the function accordingly so it can be used for both cases.

Reply:
My understanding is that this was a mistake. `MPIParallelizedStrategy` checks which subdomains are "similar" and divides the configurations to check between them. By "similar", it means which have similar _local_ homogeneities and max densities, as calculated through this - so no MPI communication needs to be done because this function should just calculate local quantities*. By including this `MPI_AllReduce` call, the function got the total numbers of particles in the entire domain. This was then used to determine bin size (more particles = smaller bins), but only local particles were considered for homogeneity and max density. - so the more MPI ranks there are, the smaller the bins. This doesn't make sense.

*there was an exception to this - in md-flexible, it appeared to be the case that the total global domain's homogeneity + max density was calculated. But this didn't work because the function was still only interesting over local particles + it was used in a misleading manner. So this was removed.

Reply:
This makes sense. Thank you!